### PR TITLE
Added audience cache

### LINKED ISSUE #22: Add support for local audience caching.
## Summary

EdgeKit audience definitions can also arrive from the server, this means to lessen the burden on servers we can cache definitions locally, and just let the server know which audience definitions we have stored locally.

### Requirements
- audience definitions should contain a flag, if they are to be cached locally
- edgekit would first check local cache, and send along some meta to the server
- server can reply with any new audience definitions or return an empty response
- the rest of the evaluation would run as normal.

### LINKED ISSUE #24: Add support for audience versions.
## Summary

Related to: #22 

This issue adds supports for audience definitions versions, which is primarily aimed at dynamic audience definitions which will evolve rapidly. We would send cached audience definition versions to the server, where a call would be made if they should be updated.

### Requirements:
- add a version field to the audience definitions
- store which audience definition version a user was matched against
- if audience definition is incremented, the user should be **un-matched** and checked again against the new version

### PR DESCRIPTION
**Pull request by k11kirky**
relies on #23 

closes #24 
closes #22 

This is an additional store that exposes the following methods:
```typescript
setAudienceCache(audiencesToCache: AudienceDefinition[])
getAudienceCache()
getAudienceCacheMetaData()
```

### CODE REVIEW

File: src/store/audienceCache.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { StorageKeys, AudienceDefinition, CachedAudienceMetaData } from 'types';
```
Comment (ydennisy) [inline src/store/audienceCache.ts:1]: I have had to remove the alias to types for now, please use a relative import.

File: src/store/audienceCache.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    this._load();
+  }
+
+  _load() {
+    const audiences: AudienceDefinition[] =
+      storage.get(StorageKeys.CACHED_AUDIENCES) || [];
+    const audiencesMetaData: CachedAudienceMetaData =
+      storage.get(StorageKeys.CACHED_AUDIENCE_META_DATA) || [];
+    this.cachedAudiences = audiences;
+    this.cachedAudiencesMetaData = audiencesMetaData;
```
Comment (ydennisy) [inline src/store/audienceCache.ts:23]: This overwrites the object you defined in the constructor, if nothing is in the store with `[]`

File: src/store/index.ts [hunk 3]
```diff
@@ -1,2 +1,3 @@
 export { audienceStore } from './audience';
 export { viewStore } from './pageview';
+export { audienceCache } from './audienceCache';
```
Comment (ydennisy) [inline src/store/index.ts:3]: We should update the naming, it is not clear the difference between a cache and a store.

Also should this be a new store/cache? (PS I know I said it should on our call)

File: src/store/audienceCache.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import { storage, timeStampInSecs } from '../utils';
+
+class AudienceCache {
+  cachedAudiences: AudienceDefinition[];
+  cachedAudiencesMetaData: CachedAudienceMetaData;
+
+  constructor() {
+    this.cachedAudiences = [];
+    this.cachedAudiencesMetaData = {
+      timeChecked: 0,
```
Comment (ydennisy) [inline src/store/audienceCache.ts:11]: timeChecked, sounds like this is related to matching / checking - why not `storedAt` or `cachedAt` this follows some of the other naming patterns.

File: src/store/index.ts (hunk 3 above)
Comment (ydennisy) [reply src/store/index.ts]: @k11kirky you did not answer my question here.

Comment (k11kirky) [reply src/store/index.ts]: @ydennisy Apologies, I think it is best to have it seperate for now, easier to test and read, but maybe a little more duplicate code. I have renamed the audience store in the lastest commits though

**Comment by ydennisy** [comment]
@k11kirky I think to get annouce LGTM status :) we still need:
- `checkedAt` to be renamed to `cachedAt`
- I feel more consistency in the naming, is possible:

```javascript
// now

export { matchedAudienceStore } from './matchedAudiences';
export { audienceCache } from './audienceCache';

// possible?
export { matchedAudienceStore } from './matchedAudiences';
export { cachedAudienceStore } from './cachedAudiences';
```

- no need for getters over straight property access, the getters are not doing anything.

- adding a new audience to cache should not overwrite all the current audiences, since some are cached, we will tell the server which can send us a delta, only new audiences. We should be able to add these new audiences, and if overwriting a check on versions also makes sense.

File: src/index.ts [hunk 5]
```diff
@@ -57,3 +61,4 @@ export const edkt = {
 
 // This will expose the exported audiences & allow tree shaking
 export * from './audiences';
+export * from './store';
```
Comment (ydennisy) [inline src/index.ts:64]: Why are we exporting `*` from store here?

File: src/store/cachedAudiences.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+          id: audience.id,
+          version: audience.version,
+        };
+      }),
+    };
+    this._save();
+  }
+
+  updateAudienceCache(audiencesToCache: AudienceDefinition[]) {
+    audiencesToCache.map((audience) => {
```
Comment (ydennisy) [inline src/store/cachedAudiences.ts:57]: This feels like a job for `forEach` as `map` is designed to return a new array.

**Comment by ydennisy** [comment]
@k11kirky there are quite a few conflicts.

File: src/index.ts (hunk 5 above)
Comment (k11kirky) [reply src/index.ts:64]: So consumers of edkt like the wrapper have access to the stores for caching, matched audiences & page views

**Pull request merged**